data-prepper-plugins/sqs-source/src/main/java/org/opensearch/dataprepper/plugins/source/sqs/BulkSqsMessageHandler.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.plugins.source.sqs;
+
+import org.opensearch.dataprepper.model.acknowledgements.AcknowledgementSet;
+import org.opensearch.dataprepper.model.buffer.Buffer;
+import org.opensearch.dataprepper.model.codec.InputCodec;
+import org.opensearch.dataprepper.model.event.Event;
+import org.opensearch.dataprepper.model.event.EventMetadata;
+import org.opensearch.dataprepper.model.record.Record;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import software.amazon.awssdk.services.sqs.model.Message;
+import software.amazon.awssdk.services.sqs.model.MessageAttributeValue;
+import software.amazon.awssdk.services.sqs.model.MessageSystemAttributeName;
+
+import java.io.ByteArrayInputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.Map;
+import java.util.concurrent.TimeoutException;
+
+public class BulkSqsMessageHandler implements SqsMessageHandler {
+    private static final Logger LOG = LoggerFactory.getLogger(BulkSqsMessageHandler.class);
+    private final InputCodec codec;
+
+    public BulkSqsMessageHandler(final InputCodec codec) {
+        this.codec = codec;
+    }
+
+    @Override
+    public void handleMessage(final Message message,
+                              final String url,
+                              final Buffer<Record<Event>> buffer,
+                              final int bufferTimeoutMillis,
+                              final AcknowledgementSet acknowledgementSet) {
+        try {
+            final String sqsBody = message.body();
+            ByteArrayInputStream inputStream = new ByteArrayInputStream(sqsBody.getBytes(StandardCharsets.UTF_8));
+            codec.parse(inputStream, record -> {
+                final Event event = record.getData();
+                final EventMetadata eventMetadata = event.getMetadata();
+                eventMetadata.setAttribute("queueUrl", url);
+                for (Map.Entry<MessageSystemAttributeName, String> entry : message.attributes().entrySet()) {
+                    final String originalKey = entry.getKey().toString();
+                    final String lowerCamelCaseKey = originalKey.substring(0, 1).toLowerCase() + originalKey.substring(1);;
+                    eventMetadata.setAttribute(lowerCamelCaseKey, entry.getValue());
+                }
+
+                for (Map.Entry<String, MessageAttributeValue> entry : message.messageAttributes().entrySet()) {
+                    final String originalKey = entry.getKey();
+                    final String lowerCamelCaseKey = originalKey.substring(0, 1).toLowerCase() + originalKey.substring(1);;
+                    eventMetadata.setAttribute(lowerCamelCaseKey, entry.getValue().stringValue());
```

Comment:
A lot of this processing can be shared, which has the added benefit of avoiding extra memory. You can do this in the refactoring that I suggest above.

That is, before the loop over the `List<Event>`, create a `Map<String, String>` for all the attributes. Then re-use those values. This should reduce both compute and memory.

Reply:
great idea, the metadata/attributes of a bulk message would remain the same for every event in that message